proposals/new-ipfs-gc.md

```diff
@@ -0,0 +1,103 @@
+It may be valuable once local named pinning is available.
+
+In the case that a user wants to remove content that was never pinned, they execute step 2 from above. This tells GC to walk the IPLD DAG identified by the CID and remove any associated blocks that are not referenced.
+
+### Impact
+While impact varies greatly for different types of users, some of our most important users (e.g. Pinata) are heavily impacted and have a need for this feature.  It will have a significant effect on their business operations.  Therefore, this gets an impact rating of 3.
+🔥🔥🔥 
+
+### Leverage
+This fixes a few specific issues, but is not highly leverageable.  The ability to remove specific content may allow implementation of a custom data retention policy.  Therefore this gets a 1 for leverage. 
+🎯
+
+### Confidence
+Medium (level 5) confidence, according to [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+## Project definition
+
+### Brief plan of attack
+
+1. Design how reference counting is implemented, exploring what interfaces will change, and what subsystems will be affected.  Address handling storage limits, preventing removal of newly arrived blocks, need for transactions in datastore, etc.
+2. Implement reference counting of blocks in the blockstore.  Counts  are updated by pinner and MFS.
+3. Implement new garbage collector that operates on reference counts instead of pinned CIDs
+4. Implement targeted GC (GC that follows a DAG and deletes associated unreferenced blocks)
+5. Optional, if it adds value: Implement command to remove pin and then immediately GC the unpinned CID
+6. Distribute an experimental version of go-ipfs for evaluation and feedback.
+
+Garbage collection performance issues arise because GC attempts to remove blocks by loading a set of all pinned CIDs into memory and searching for each block’s CID in this set.  The plan attack operates on the following premises:
+
+- Maintaining reference counts for blocks will make it quick to determine if a block can be removed, because it is not necessary to load all pinned CID into memory and search this.
+- Specific content can be removed by performing GC on the blocks associated with a IPLD DAG identified by CID.
+
+### What does done look like?
+- Block reference counts stored in datastore
+- Block reference counts updated by pinning, unpinning, adding to MFS, removing from MFS. 
+- New GC implementation that operates on block reference counts  instead of on  set of all pinned CIDs
+- Ability for GC to remove content identified by CID
+  - New `--cid` option: `ipfs repo gc --cid=<cid>`
+
+Removing pins will be slower because it requires walking DAG to decrement reference counts. Adding items to MFS and removing items from MFS will be slower due to needing to update reference counts.  Hopefully any increase in execution time will be insignificant.
```

Comment:
Seems like we should use piñata as an example here and estimate what the impact would be (since they add pins all the time, but don’t GC as often now)

Reply:
I caught up with Matt to try and grab some numbers here: 
https://www.notion.so/Pinata-5157fa6d2a4741a593bb8ff32bfdb08e

One thing that stuck out to me was the two growth rates of the platforms (one is generating 5-6k pins a day, another 5k pins a day). That coupled with GC windows taking 2x as long every 3-4 months (as of a few months ago) may not have fully captued what's happening.

Reply:
@momack2 I will add some rough estimates by timing some things that I think will be similar in time to the new GC, and comparing with the old GC. 

@jnthnvctr - I added a link to this in the doc, and brought some of the content into the doc.  Thank you!